cmd/bench/cmd/node.go

```diff
@@ -0,0 +1,182 @@
+	"github.com/filecoin-project/mir/pkg/net/grpc"
+	"github.com/filecoin-project/mir/pkg/net/libp2p"
+	"github.com/filecoin-project/mir/pkg/requestreceiver"
+	t "github.com/filecoin-project/mir/pkg/types"
+	grpctools "github.com/filecoin-project/mir/pkg/util/grpc"
+	libp2ptools "github.com/filecoin-project/mir/pkg/util/libp2p"
+)
+
+const (
+	NodeBasePort        = 10000
+	ReqReceiverBasePort = 20000
+)
+
+var (
+	transportType string
+	statFileName  string
+	statPeriod    time.Duration
+
+	nodeCmd = &cobra.Command{
+		Use:   "node",
+		Short: "Start a Mir node",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return runNode()
+		},
+	}
+)
+
+func init() {
+	rootCmd.AddCommand(nodeCmd)
+	nodeCmd.Flags().StringVarP(&transportType, "net", "t", "libp2p", "network transport (libp2p|grpc)")
+	nodeCmd.Flags().StringVarP(&statFileName, "statFile", "o", "", "output file for statistics")
+	nodeCmd.Flags().DurationVar(&statPeriod, "statPeriod", time.Second, "statistic record period")
+}
+
+func runNode() error {
+	var logger logging.Logger
+	if verbose {
+		logger = logging.ConsoleDebugLogger
+	} else {
+		logger = logging.ConsoleWarnLogger
+	}
+
+	ownID, err := strconv.Atoi(id)
+	if err != nil {
+		return fmt.Errorf("unable to convert node ID: %w", err)
+	} else if ownID < 0 || ownID >= nrNodes {
+		return fmt.Errorf("ID must be in [0, %d]", nrNodes-1)
+	}
+
+	nodeIDs := make([]t.NodeID, nrNodes)
+	for i := range nodeIDs {
+		nodeIDs[i] = t.NewNodeIDFromInt(i)
+	}
+
+	reqReceiverAddrs := make(map[t.NodeID]string)
+	for i := range nodeIDs {
+		reqReceiverAddrs[t.NewNodeIDFromInt(i)] = fmt.Sprintf("127.0.0.1:%d", ReqReceiverBasePort+i)
+	}
+
+	var transport net.Transport
+	nodeAddrs := make(map[t.NodeID]t.NodeAddress)
+	switch strings.ToLower(transportType) {
+	case "grpc":
+		for i := range nodeIDs {
+			nodeAddrs[t.NewNodeIDFromInt(i)] = t.NodeAddress(grpctools.NewDummyMultiaddr(i + NodeBasePort))
+		}
+		transport, err = grpc.NewTransport(t.NodeID(id), nodeAddrs[t.NodeID(id)], logger)
+	case "libp2p":
+		h := libp2ptools.NewDummyHost(ownID, NodeBasePort)
+		for i := range nodeIDs {
+			nodeAddrs[t.NewNodeIDFromInt(i)] = t.NodeAddress(libp2ptools.NewDummyMultiaddr(i, NodeBasePort))
+		}
+		transport, err = libp2p.NewTransport(h, t.NodeID(id), logger)
+	default:
+		return fmt.Errorf("unknown network transport %s", strings.ToLower(transportType))
+	}
+	if err != nil {
+		return fmt.Errorf("failed to get network transport %w", err)
+	}
+
+	issConfig := iss.DefaultConfig(nodeIDs)
+	issProtocol, err := iss.New(t.NodeID(id), issConfig, logger)
+	if err != nil {
+		return fmt.Errorf("could not instantiate ISS protocol module: %w", err)
+	}
+
+	// Use dummy crypto module that only produces signatures
+	// consisting of a single zero byte and treats those signatures as valid.
+	// TODO: Adjust once a default crypto implementation is provided by Mir.
+	crypto := mirCrypto.New(&mirCrypto.DummyCrypto{DummySig: []byte{0}})
```

Comment:
What exactly do we want to measure with this benchmark?
If it is the performance of ISS, then crypto (I guess) is an important part and you cannot use dummy crypto.

Reply:
@matejpavlovic Do we have a better Crypto module implementation to banchmark with?

Reply:
Yes, there is a ["Pseudo" crypto implementation](https://github.com/filecoin-project/mir/blob/main/pkg/crypto/pseudo.go#L29) for this purpose. It actually performs all the cryptographic operations, just the keys are generated deterministically, so it's easier to set up. You just need to give it the node IDs on instantiation. [Here's an example](https://github.com/filecoin-project/mir/blob/main/pkg/iss/iss_test.go#L371) of how it's used.